WKWebView's scrollView.bounces=false is not working

nmoucht opened the pull request:
#### b40f80bcf189b67147eb79d9e0ca3c00dc8f8ea0
<pre>
WKWebView&apos;s scrollView.bounces=false is not working
<a href="https://bugs.webkit.org/show_bug.cgi?id=243270">https://bugs.webkit.org/show_bug.cgi?id=243270</a>
&lt;rdar://96024524&gt;

Reviewed by Simon Fraser.

Keep track if bounces is set by client to make sure it is not overwritten by overscroll behavior.
Currently setting bounces = True on a page with overscroll-behavior:none will result in the client
setting to be ignored.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::updateScrollViewForOverscrollBehavior):

Canonical link: <a href="https://commits.webkit.org/254582@main">https://commits.webkit.org/254582@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/401e301ae59fa5c763ebbccb2fc694ba98ffabb6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89485 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34039 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20218 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98798 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155105 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93493 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32540 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81843 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93211 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95132 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25850 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76376 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25784 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80728 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80622 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68775 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30299 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14696 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30041 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15633 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3221 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33488 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38589 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32198 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34724 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
```diff
@@ -223,8 +223,8 @@ - (void)cancelPointersForGestureRecognizer:(UIGestureRecognizer *)gestureRecogni
-    scrollView.bouncesHorizontally = horizontalOverscrollBehavior != OverscrollBehavior::None;
-    scrollView.bouncesVertically = verticalOverscrollBehavior != OverscrollBehavior::None;
+    scrollView.bouncesHorizontally &= horizontalOverscrollBehavior != OverscrollBehavior::None;
+    scrollView.bouncesVertically &= verticalOverscrollBehavior != OverscrollBehavior::None;
```
This isn't the right way to fix the bug.

It needs to be fixed such that a webkit client can read and write the value of `scrollView.bounces` and get consistent behavior (i.e. they get back the value they set). This needs to work even when webkit internally sets the value because of OverscrollBehavior.

The way we do this is to put code in WKScrollView to track the internal state; see `_setScrollEnabledInternal` etc.

There will be API tests that verify this behavior, and you should make one for this case too.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
```diff
@@ -265,6 +263,8 @@ - (void)cancelPointersForGestureRecognizer:(UIGestureRecognizer *)gestureRecogni
         }
         if (scrollingStateNode.hasChangedProperty(ScrollingStateNode::Property::ScrollableAreaParams)) {
             auto params = scrollingStateNode.scrollableAreaParameters();
+            scrollView.bouncesHorizontally = params.horizontalOverscrollBehavior != OverscrollBehavior::None;
```
Shouldn't this be calling `_setBouncesInternal`?

nmoucht replied:
This code is only called for scrolling nodes, not the main document. Is it necessary to keep track of this state even if the client cannot set its value?

smfr replied:
Ah you're right. You undid some code sharing in this patch, which is unfortunate. I think it would be better to continue to share code, but know whether we're updating the main UIScrollView or not.

nmoucht replied:
Good idea. Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
```diff
@@ -221,10 +222,14 @@ - (void)cancelPointersForGestureRecognizer:(UIGestureRecognizer *)gestureRecogni
 {
-    scrollView.bouncesHorizontally = horizontalOverscrollBehavior != OverscrollBehavior::None;
-    scrollView.bouncesVertically = verticalOverscrollBehavior != OverscrollBehavior::None;
+    if (isWKScrollView == IsMainUIScrollView::Yes)
```
Instead of passing in `IsMainUIScrollView` you could just do:
```
if ([scrollView isKindOfClass:[WKScrollView class]]) ...
```

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/ios/WKScrollView.mm:
```diff
@@ -464,6 +470,25 @@ - (void)_updateScrollability
 
+- (void)setBounces:(BOOL)value
+{
+    _bouncesSetByClient = value;
```
I know this follows existing naming, but I find that naming confusing. There's ambiguity about whether this is "the value set by the client" or "the client has set some value". "set by client" implies the second. I'd prefer we rename them all to _clientBouncesValue or something.

smfr replied:
It would be clearer if these were std::optional<> too.

nmoucht replied:
Agreed. This probably warrants a follow up patch.

smfr commented on Source/WebKit/UIProcess/ios/WKScrollView.mm:
```diff
@@ -464,6 +470,25 @@ - (void)_updateScrollability
+
+- (void)_updateBouncability
+{
+    [super setBouncesHorizontally:(_bouncesSetByClient && _bouncesHorizontalInternal)];
```
This means that a client can only turn a bounces=YES into a bounces=NO, but should they be allowed to enable bouncing on a page with `overscroll-behavior:none`? I don't know. But since you made a decision here, document it in the commit message.

nmoucht replied:
Added to the commit message. I agree that it's unclear what the right behavior here should be. For the follow up patch I'll try and see what other properties are doing and make them consistent.

smfr commented on Tools/TestWebKitAPI/Tests/ios/ScrollViewBouncesTests.mm:
```diff
@@ -0,0 +1,191 @@
+{
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 320, viewHeight)]);
+    [webView synchronouslyLoadHTMLString:overscrollBehaviorAuto];
+    [[webView scrollView] setBounces:NO];
```
How about a test that calls -setBounces: BEFORE the load?

You could also test that a page that dynamically changes `overscroll-behavior` does the right thing.

nmoucht replied:
Fixed.

nmoucht committed:
WKWebView's scrollView.bounces=false is not working
https://bugs.webkit.org/show_bug.cgi?id=243270
<rdar://96024524>

Reviewed by Simon Fraser.

Keep track if bounces is set by client to make sure it is not overwritten by overscroll behavior.
Currently setting bounces = True on a page with overscroll-behavior:none will result in the client
setting to be ignored.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::updateScrollViewForOverscrollBehavior):

Canonical link: https://commits.webkit.org/254582@main
